presto-main/src/main/java/io/prestosql/sql/planner/EffectivePredicateExtractor.java

```diff
@@ -336,69 +340,103 @@ public Expression visitValues(ValuesNode node, Void context)
                 return TRUE_LITERAL;
             }
 
-            // get all types in one shot -- needed for the expression optimizer below
-            List<Expression> allExpressions = node.getRows().stream()
-                    .flatMap(List::stream)
-                    .collect(Collectors.toList());
-
-            Map<NodeRef<Expression>, Type> expressionTypes = typeAnalyzer.getTypes(session, types, allExpressions);
-
-            ImmutableMap.Builder<Symbol, Domain> domains = ImmutableMap.builder();
-
-            for (int column = 0; column < node.getOutputSymbols().size(); column++) {
-                Symbol symbol = node.getOutputSymbols().get(column);
-                Type type = types.get(symbol);
+            // for each row of Values, get all expressions that will be evaluated:
+            // - if the row is of type Row, evaluate fields of the row
+            // - otherwise evaluate the whole expression and then analyze fields of the resulting row
+            checkState(node.getRows().isPresent(), "rows is empty");
+            List<Expression> processedExpressions = node.getRows().get().stream()
+                    .flatMap(row -> {
+                        if (row instanceof Row) {
+                            return ((Row) row).getItems().stream();
+                        }
```

Comment:
I wanted to take advantage of analyzing Row items separately in case some of them are non-deterministic and other are deterministic. When analyzing them one by one, we still can build domains for the deterministic columns.